Add FastAPI routes for tours

jdavcs opened the pull request:
Ref #10889 

@mvdbeek, @davelopez: please take a look - here's my first attempt.

Note: A pydantic `TourList` model (as opposed to `List[Tour]`) was necessary for correct serialization in `web/framework/decorators.py`. Even though we can modify [this logic](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/framework/decorators.py#L342) to recognize a list of pydantic models, such a list cannot be serialized using the `[model].json()` method.

jdavcs committed:
Add missing id keys to built-in tours

Pydantic models (currently added to tours) require explicit
specification of model fields. Having an "id" field as required seems
right from a design standpoint.

However, if there is a reason for an "id" value to be included in the
galaxy_ui tour but not in the other tours, this field can be made
optional: `id: Optional[str] = None`

jdavcs committed:
Add test, assertions for tour API testing

jdavcs committed:
Add FastAPI routes, Pydantic models to tours

jdavcs committed:
Make a TourList model for correct serialization

jdavcs committed:
Refactor tour loading (minor)

- Do not call tour_loader() twice
- Do not return a value from a function that modifies the value of its
  argument in place
- Rename function to reflect what it does

jdavcs committed:
Simplify parsing pydantic models

jdavcs committed:
Move pydantic models to schema/

davelopez commented on lib/galaxy/webapps/galaxy/api/tours.py:
```diff
@@ -3,16 +3,52 @@
+
+@cbv(router)
+class FastAPITours:
+    app: UniverseApplication = Depends(get_app)
```
Maybe something more specific here like:
```suggestion
    tour_registry: Registry = Depends(get_tour_registry)
```

where `get_tours_registry` is:

`````python
def get_tour_registry(app: UniverseApplication = Depends(get_app)) -> Registry:
    return app.tour_registry
````

jdavcs replied:
Done.

davelopez commented on lib/galaxy/schema/__init__.py:
```diff
@@ -1,13 +1,44 @@
+    id: str
+    name: str
+    description: str
+    tags: List[str]
```
Will the documentation of the fields (for the OpenAPI docs) be added in a separate PR?

jdavcs replied:
@davelopez I can add it now. What are the required fields? (I'm looking at OpenAPI specs..) Is it just title, description, example?

davelopez replied:
Yes, those seem to be the most used for documentation, you can also define some restrictions about the fields with the other properties (`ge`, `max_length`, etc.) if it is required.

jdavcs replied:
Done.

mvdbeek commented on lib/galaxy/schema/__init__.py:
```diff
@@ -1,13 +1,44 @@
+
+
+class TourStep(BaseModel):
+    title: Optional[str] = None
```
Can you add a description to these fields ?
Something like
```suggestion
    title: Optional[str] = Field(title="Title", description="The title of this tour step", example="Click here")
```
?

jdavcs replied:
Done.

davelopez commented on lib/galaxy/schema/__init__.py:
```diff
@@ -1,13 +1,44 @@
         return []
+
+
+class Tour(BaseModel):
```
Should we put all the API BaseModels for all the routes under `schema/__init__.py` or should they be in their own file like `schema/tours.py`? I am still struggling with Python packages/modules/__init__ best practices 😅

mvdbeek replied:
They should be with the tours, since we won't re-use the model in other places, so lib/galaxy/tours/schema.py seems like a good idea.

jdavcs replied:
Done! `tours/schema.py`.

mvdbeek commented on lib/galaxy/schema/__init__.py:
```diff
@@ -1,13 +1,44 @@
         return []
+
+
+class Tour(BaseModel):
```
Since those are probably only used with tours I'd probably put this with `lib/galaxy/tours` ?

jdavcs replied:
That was my first version - until the last PR.
Moving to `tours/schema.py` (as per discussion below)

davelopez reviewed (commented):
Looking great!

davelopez commented on lib/galaxy/tours/__init__.py:
```diff
@@ -23,20 +26,24 @@ def tour_loader(contents_dict):
+                'tags': self.tours[k].get('tags')
+            }
+            tours.append(tourdata)
+        return parse_obj_as(TourList, tours)
```
I think you can use here:
```suggestion
        return parse_obj_as(List[Tour], tours)
```
In case you want to use the `List` typing instead of a model for the list. But not 100% sure :)

jdavcs replied:
Unless I'm mistaken, we can't. That was my first approach, but I ran into the issue I mention in the PR description: this has to be serialized to json, and we cannot use pydantic's `model.json()` if it's a list of pydantic models: we can only serialize each model. To serialize the list you need to use `json.dumps()` (or our `safe_dumps` - same thing) - but `json.dumps()` doesn't know how to handle a model. Thus, using a pydantic model which is a list of models seems to be the most straightforward way to implement this. (It's different for the licenses endpoint where the response model is a list because we don't serialize that list of models, at least for now)

jmchilton commented on config/plugins/tours/core.history.yaml:
```diff
@@ -1,3 +1,4 @@
+id: galaxy_history
```
Can you explain adding the id to the tours here? Does this get propagated somehow? It looks like the id was calculated in a different way before?

jdavcs replied:
The id is calculated based on the filename [here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tours/__init__.py#L73), so the id from the yaml file is not used at all for tour loading. I checked the client ([client/src/mvc/tours.js](https://github.com/galaxyproject/galaxy/blob/dev/client/src/mvc/tours.js#L43) appears to be the only relevant file) - same there. I considered deleting the id field, but instead decided to add id values to the other 2 tours for consistency, in case there may have been plans for that field. I don't mind either deleting the field or reverting my edit and leaving as is, but having that field optional (given that we formalize it in a pydantic model) doesn't seem right, I think.

jmchilton replied:
Thanks for looking into that - so it seems it isn't used in a functional capacity anywhere? Want to remove it from core.galaxy_ui.yaml and see if anyone complains? I think we should remove it if it is not being used.

jdavcs replied:
Done! Completely agree on removing dead code.

jdavcs committed:
Move tour pydantic models to their own module

jdavcs committed:
Remove id field from tour yaml files

Field not used. See code review discussion here:
https://github.com/galaxyproject/galaxy/pull/11089#discussion_r555072412

jdavcs committed:
Add model field metadata; revise tour model

Address comments in code review
Model is revised due to change in Tour model spec:
- Id field is no longer part of the yaml file spec;
- However, an id is generated and added to the tour data returned as a
list
- As a result, in a list, a tour item must have an id, whereas in tour
details, a tour item must not have an id.
I've factored out "core tour" data into its own model to solve this.

jdavcs committed:
Use a registry abc instead of concrete class

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -85,3 +86,7 @@ def get_admin_user(trans: SessionRequestContext = Depends(get_trans)):
     return trans.user
+
+
+def get_tours_registry(app: UniverseApplication = Depends(get_app)) -> ToursRegistry:
```
Not sure, maybe we should move this function to `lib/galaxy/webapps/galaxy/api/tours.py`?. It seems probable that it will be used only in the `tours` route.

jdavcs replied:
Done. I didn't include the `UniverseApplication` annotation to avoid importing it into `api/tours.py` (it's already annotated in `get_app`)

davelopez reviewed (commented):
Cool! I like your approach of using an ABC for the ToursRegistry :)

jdavcs committed:
Refactor tours package organization

jdavcs committed:
Refactor tours module implemantation

jdavcs committed:
Work around limitations of mypy

jmchilton commented on lib/galaxy/tours/__init__.py:
```diff
@@ -1,87 +1,7 @@
+from ._schema import TourCore  # noqa
+from ._schema import TourStep  # noqa
+from ._schema import TourDetails  # noqa
+from ._schema import TourList  # noqa
```
Instead of ``#noqa`` please use ``__all___`` at the end of the file.

```
__all__ = ("build_tours_registry", "Tour", ...)
```

jdavcs replied:
Done

jmchilton replied:
Thanks!

jdavcs committed:
Define __all__ list for tours package

jmchilton reviewed (approved):
Amazing! Thanks for the deep dive into this.
